### ISSUE TITLE
NonUniformImage wrong image when using large values for axis

### ISSUE THREAD

**Issue by funkygoby** [matplotlib/matplotlib#15039]
When using `matplotlib.image.NonUniformImage` with large values on axis (~10e+7), the plot breaks.

I am plotting a random matrix and simply offsetting the axis by with a large value:

**Normal behaviour**
![ok](https://user-images.githubusercontent.com/54024979/62860138-7eb53d00-bcff-11e9-98d0-0d92bc0e1e47.png)

**Strange behaviour**
![not_ok](https://user-images.githubusercontent.com/54024979/62860182-a0162900-bcff-11e9-90f2-d433c0ec2786.png)

**The code**
```
import numpy
import matplotlib.pyplot as plt
from matplotlib.image import NonUniformImage

offset = 9.*10**6 # XXX change this XXX
nx = 60
ny = 100
x = numpy.arange(offset, offset+nx)
y = numpy.arange(offset, offset+ny)
z = numpy.random.rand(ny, nx)

ax = plt.subplot(111)
im = NonUniformImage(ax, extent=(offset, offset+nx, offset, offset+ny))
im.set_data(x, y, z)
ax.images.append(im)
ax.set_xlim(offset, offset+nx)
ax.set_ylim(offset, offset+ny)

plt.show()
```

By changing the `offset` variable, one can achieve different results: `1.` is ok; `9.` is broken.

Thank you

**Comment by timhoffm** [comment]
Confirmed on master.

Observation: The result depends on the figsize:

![grafik](https://user-images.githubusercontent.com/2836374/63716991-8e786980-c847-11e9-874a-ce08933ec33d.png)
![grafik](https://user-images.githubusercontent.com/2836374/63717024-a05a0c80-c847-11e9-95e0-87e539449f59.png)
![grafik](https://user-images.githubusercontent.com/2836374/63717088-b9fb5400-c847-11e9-9417-206f583f88bd.png)
![grafik](https://user-images.githubusercontent.com/2836374/63717122-d0091480-c847-11e9-93cf-4f6c132a14f0.png)
![grafik](https://user-images.githubusercontent.com/2836374/63717158-e1522100-c847-11e9-8842-34741c4235ea.png)

Probably there is some kind of coordinate to data index mapping problem.

**Comment by anntzer** [comment]
This appears to be fixed by #14913 (and it's not "just" a naive precision problem -- replacing floats by doubles everywhere in the c++ code doesn't seem enough to fix it.
